flow/stopwatch_dl.cc

```diff
@@ -0,0 +1,151 @@
+                                        /*right=*/bar_left + bar_width,
+                                        /*bottom=*/bottom),
+                       0xAA0000FF);
+    }
+  }
+
+  // Draw horizontal frame markers.
+  {
+    if (max_interval > one_frame_ms) {
+      // Paint the horizontal markers.
+      auto count = static_cast<size_t>(max_interval / one_frame_ms);
+
+      // Limit the number of markers to a reasonable amount.
+      if (count > kMaxFrameMarkers) {
+        count = 1;
+      }
+
+      for (auto i = size_t(0); i < count; i++) {
+        auto const frame_height =
+            height * (1.0 - (UnitFrameInterval(i + 1) * one_frame_ms) /
+                                max_unit_interval);
+
+        // Draw a skinny rectangle (i.e. a line).
+        painter.DrawRect(SkRect::MakeLTRB(/*left=*/x,
+                                          /*top=*/y + frame_height,
+                                          /*right=*/width,
+                                          /*bottom=*/y + frame_height + 1),
+                         0xCC000000);
+      }
+    }
+  }
+
+  // Paint the vertical marker for the current frame.
+  {
+    DlColor color = DlColor::kGreen();
+    if (UnitFrameInterval(stopwatch_.LastLap().ToMillisecondsF()) > 1.0) {
+      // budget exceeded.
+      color = DlColor::kRed();
+    }
+    auto const l =
+        x + width * (static_cast<double>(stopwatch_.GetCurrentSample()) /
+                     kMaxSamples);
+    auto const t = y;
+    auto const r = l + width * sample_unit_width;
+    auto const b = rect.bottom();
+    painter.DrawRect(SkRect::MakeLTRB(l, t, r, b), color);
+  }
+
+  // Actually draw.
+  canvas->DrawVertices(painter.IntoVertices(), DlBlendMode::kSrc, paint);
+}
+
+void DlVertexPainter::DrawRect(const SkRect& rect, const DlColor& color) {
+  // Draw 6 vertices representing 2 triangles.
+  auto const left = rect.x();
+  auto const top = rect.y();
+  auto const right = rect.right();
+  auto const bottom = rect.bottom();
+
+  auto const vertices = std::array<SkPoint, 6>{
+      SkPoint::Make(left, top),      // tl tr
+      SkPoint::Make(right, top),     //    br
+      SkPoint::Make(right, bottom),  //
+      SkPoint::Make(right, bottom),  // tl
+      SkPoint::Make(left, bottom),   // bl br
+      SkPoint::Make(left, top)       //
+  };
+
+  auto const colors = std::array<DlColor, 6>{
+      color,  // tl tr
+      color,  //    br
+      color,  //
+      color,  // tl
+      color,  // bl br
+      color   //
+  };
+
+  vertices_.insert(vertices_.end(), vertices.begin(), vertices.end());
+  colors_.insert(colors_.end(), colors.begin(), colors.end());
+}
+
+std::shared_ptr<DlVertices> DlVertexPainter::IntoVertices() {
+  auto const result = DlVertices::Make(
+      /*mode=*/DlVertexMode::kTriangles,
+      /*vertex_count=*/vertices_.size(),
+      /*vertices=*/vertices_.data(),
+      /*texture_coordinates=*/nullptr,
+      /*colors=*/colors_.data());
+  vertices_.clear();
+  colors_.clear();
```

Comment:
So I'm not sure if this is actually valud. Does the DlVertices constructor copy the data out, or is it being stored by reference? If you clear the data here won't the vertices buffers end up filled with junk?

Reply:
Looks like it does copy: https://github.com/flutter/engine/blob/main/display_list/dl_vertices.cc#L39

So this is fine!

Reply:
A good question.

I did write a unit test verifying that it doesn't _appear_ to be junk. My understanding is it's probably copied (it's just `SkPoint` and `DlColor`), but maybe I shouldn't expect that to always be true.

I could either:

1. Explore moving the data (which probably is better given the name `IntoVertices()`?)
2. Copy the data explicitly before passing it
3. Do nothing, i.e. just mark with a boolean that this class is no longer valid and `FML_DCHECK`.

I'd prefer to do 1 > 2 > 3, wdut though?

Reply:
I don't think any action is necessary.

Reply:
Ack, thanks for the discussion though!